### ISSUE TITLE
Login fails with `security.enableWrappers = false`

### ISSUE THREAD

**Issue by smaeul** [NixOS/nixpkgs#369441]
## Describe the bug

With `security.enableWrappers = false`, login fails because the `unix_chkpwd` binary is not found in the expected location:

```
pam_unix(login:account): helper binary execve failed: No such file or directory
```

PAM expects to find the helper in the wrapper directory, which does not exist:

```
$ grep -z /unix_chkpwd /nix/store/bxk2n72mwfb6jz9ppywqz2d0f4s3lxha-linux-pam-1.6.1/lib/security/pam_unix.so
/run/wrappers/bin/unix_chkpwd
```

## Steps To Reproduce

1. Build NixOS with `security.enableWrappers = false`.
2. Attempt to log in (e.g. set `services.getty.autologinUser = root`).
3. Login fails with the above error message printed in the journal.

## Expected behavior

When `security.enableWrappers = false`, the [linux-pam patch pointing to the wrapper directory](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/li/linux-pam/suid-wrapper-path.patch) should be excluded.

## Additional context

I am building a minimal system that does not need any form of privilege escalation, so it is fine if `unix_chkpwd` is not setuid. But it seems that several packages assume `security.enableWrappers = true`.

## Metadata

The affected system is an appliance that does not include the nix package manager. Here is the metadata from the build system (same architecture, and built from the same flake, so the same nixpkgs version):

 - system: `"aarch64-linux"`
 - host os: `Linux 6.6.64, NixOS, 24.11 (Vicuna), 24.11.20241213.314e12b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/nix/store/0xbni69flk8380w0apw4h640n37wn1i9-source`

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by lucasew** [comment]
It seems like it's expected because these wrappers are SUID wrappers. SUID stuff in other distros would be stored in /sbin but as NixOS is not a FHS distro this wrapping scheme is required.

That's why running sudo from /run/current-system/sw/bin for example doesn't work.

**Comment by smaeul** [comment]
In my case, the lack of SUID is not a problem, because `login(1)` is running as root. So the system works correctly if the patch is not applied and PAM tries to execute the unwrapped path (`/nix/store/bxk2n72mwfb6jz9ppywqz2d0f4s3lxha-linux-pam-1.6.1/bin/unix_chkpwd`) instead.

My system does not have sudo or equivalent installed, so it is fine that those wouldn't work.

**Comment by trofi** [comment]
`libpam` is a popular dependency. You would need to rebuild all reverse dependencies that lead up to `login` (at least `shadow` itself) if the patch to be made conditional.

I see a few options here:
1. you can build `shadow` without `pam` support: untested `shadow = prev.shadow.override { pam = null; }` overlay could do the trick
2. `nixpkgs` could expose a knob for `libpam` to make the patch application optional. That way an overlay with `libpam = prev.libpam.override { withNixosWrapper = false; }`. could do the trick.
3. something else

It feels like none of those would be reasonable for `security.enableWrappers = false;` knob. It could be easier to provide `/run/wrappers/bin/unix_chkpwd` as a symlinks to the binary directly.

**Comment by Atemu** [comment]
Is there a reason we don't just take `unix_chkpwd` from `$PATH`? That way it'd work both ways without modifications necessary.

**Comment by trofi** [comment]
`libpam` can be used by an unprivileged process to validate the password of a current user. A typical example is a lock screen, like `i3lock`. To do that `libpam` calls `unix_chkpwd` SUID to read `/etc/shadow`. Dropping SUID will break lockscreen unlocking.

**Comment by trofi** [comment]
On `pam`-less systems the lockscreen itself would have to be SUID-marked to work.

**Comment by smaeul** [comment]
> `libpam` is a popular dependency. You would need to rebuild all reverse dependencies that lead up to `login` (at least `shadow` itself) if the patch to be made conditional.

Yeah, this is why I went back to `security.enableWrappers = true`.

> I see a few options here:
> 
> 1. you can build `shadow` without `pam` support: untested `shadow = prev.shadow.override { pam = null; }` overlay could do the trick

I was hesitant to do this because a PAM-less system seemed like it might be even less supported.

> 2. `nixpkgs` could expose a knob for `libpam` to make the patch application optional. That way an overlay with `libpam = prev.libpam.override { withNixosWrapper = false; }`. could do the trick.

This solution would require a similar knob for every package that provides a SUID-wrapped binary. A couple of them already do this, presumably to support building outside NixOS, but most do not. I'm not very familiar with how the modules and packages are allowed to interact with each other, but I'm guessing there's no way to check `security.enableWrappers` in the package, so the user would need to set all of these knobs manually?

> 3. something else
> 
> 
> It feels like none of those would be reasonable for `security.enableWrappers = false;` knob. It could be easier to provide `/run/wrappers/bin/unix_chkpwd` as a symlinks to the binary directly.

Yes, this certainly seems like a simpler solution. If all of the wrapped binaries were otherwise in `$PATH`, we could symlink `/run/wrappers/bin` to `/run/current-system/sw/bin` and ignore `security.wrappers`. But this is not the case (e.g. `dbus-daemon-launch-helper`), so each item in `security.wrappers` needs its own symlink.

And then it doesn't solve my original motivation for turning off the wrappers, which was removing FUSE from the system closure. I'd still need something like my current hack, which is:

```nix
security.wrappers = {
  fusermount.source = lib.mkForce "${pkgs.coreutils}/bin/false";
  fusermount3.source = lib.mkForce "${pkgs.coreutils}/bin/false";
};
```

**Comment by Atemu** [comment]
You could perhaps try `security.wrappers.fusermount = lib.mkForce { }`. If that doesn't work you could add an enable option to the submodule and filter based on it.

**Comment by zackw** [comment]
N.B. there is now a top-level off switch for FUSE, `programs.fuse.enable = false`.
